consensus/safety-rules/src/persistent_safety_storage.rs

```diff
@@ -109,14 +113,17 @@ impl PersistentSafetyStorage {
     }
 
     pub fn safety_data(&self) -> Result<SafetyData, Error> {
-        Ok(self.internal_store.get(SAFETY_DATA).map(|v| v.value)?)
+        Ok(self.cached_safety_data.clone())
     }
 
     pub fn set_safety_data(&mut self, data: SafetyData) -> Result<(), Error> {
         counters::set_state("epoch", data.epoch as i64);
         counters::set_state("last_voted_round", data.last_voted_round as i64);
         counters::set_state("preferred_round", data.preferred_round as i64);
-        self.internal_store.set(SAFETY_DATA, data)?;
+
+        self.internal_store.set(SAFETY_DATA, data.clone())?;
```

Comment:
Once we have check-and-set with Vault, this needs to invalidate the cache if the write fails so that it gets re-read the next time.

Reply:
I think we should be clear what the right behavior is... in general, if safety rules ever hits an invariant such as cache failing, we should let the admin know even if we automatically recover. this could be because bad things happened...

Reply:
Currently this will happen on upgrades with a new genesis (because the vault-init TF resets the safety_data in Vault). In production it should never happen though. In my mind the model is that whatever is in Vault is canonical, but I can see a model where the LSR process is also canonical and if its view of the world doesn't match what's in Vault it just stops processing until the operator intervenes.

Reply:
What I've done for now is throw an error if the write fails, which as I understand it, LSR will log but continue execution (similar to any write error as it stands today). I'm not sure we want to panic (we try to avoid doing that in LSR as a matter of principle), so I've handled this case as any other storage failure-related case: if the write fails (for any reason), invalidate the local cache and log the error.

Reply:
Yup, this is the right behaviour.

Reply:
If the write fails, it will fail the current vote though right? (i.e. it won't return the vote to the validator)

Reply:
Yeah, it'll behave the same as any other failure that's thrown during execution (i.e., exit early and return the error to the caller -- which should be the safety rules client in a remote safety rules deployment).